library/Icinga/Web/Controller/ActionController.php

```diff
@@ -450,6 +451,22 @@ public function preDispatch()
             $form->handleRequest();
         }
         $this->_helper->layout()->autoRefreshForm = $form;
+
+        if (! $this->isXhr()
+            // Layout is only enabled for initial page loads or page reloads
+            && $this->_helper->layout()->isEnabled()
```

Comment:
In which case is it not an XHR request and without layout?

Reply:
To answer my own question: Static asset (e.g. images) requests and requests with the `_disableLayout` query parameter.

There are other responses without layout. Some of them include JSON or CSV content. Though some still include HTML.

So to clarify what's happening here, I'd suggest to check if the content type is `text/html` which it is only if this is true:

* Not a static asset request
* Not an api request
* Not an export

Though, to check the content type, the CSP header must not be set until the response is ready. i.e. as part of `postDispatch` (which won't be used in some of the cases above anyway)

Reply:
Changed it. The header is now registered by the response though. https://github.com/Icinga/icingaweb2/pull/5059/commits/b96894202402950b9c3725e5f01aad4a049f6f51